File: ios/HiveMobile/Stores/SubAgentStatus.swift
import Foundation

enum SubAgentExecutionState: Equatable {
    case pending
    case running
    case completed
    case failed
}

private let runningAgentStatuses: Set<String> = ["pendinginit", "running", "interrupted"]
private let failedAgentStatuses: Set<String> = ["errored", "notfound", "failed"]

func subAgentExecutionState(
    for tool: ToolCall,
    children: [ToolCall] = [],
    childrenByParentId: [String: [ToolCall]] = [:],
    showExecutingState: Bool = false
) -> SubAgentExecutionState {
    guard tool.name == "Task" || tool.name == "Agent" else {
        return tool.output == nil ? .pending : .completed
    }

    let inputPayload = jsonObject(from: tool.input)
    let outputPayload = tool.output.flatMap(toolOutputPayload)
    let toolStatus = stringValue(outputPayload?["status"] ?? inputPayload?["status"]).map(normalizeStatus)
    let agentStatuses = agentStatuses(from: inputPayload) + agentStatuses(from: outputPayload)

    if toolStatus == "failed" || agentStatuses.contains(where: { failedAgentStatuses.contains($0) }) {
        return .failed
    }

    guard showExecutingState else {
        return tool.output == nil ? .pending : .completed
    }

    if tool.output == nil {
        return .running
    }

    let collabTool = stringValue(inputPayload?["tool"] ?? outputPayload?["tool"])
    if collabTool == "spawnAgent" && agentStatuses.contains(where: { runningAgentStatuses.contains($0) }) {
        return .running
    }

    if hasRunningDescendant(children, childrenByParentId: childrenByParentId) {
        return .running
    }

    return .completed
}

private func hasRunningDescendant(
    _ children: [ToolCall],
    childrenByParentId: [String: [ToolCall]]
) -> Bool {
    for child in children {
        let nestedChildren = childrenByParentId[child.id] ?? []
        if child.name == "Task" || child.name == "Agent" {
            if subAgentExecutionState(
                for: child,
                children: nestedChildren,
                childrenByParentId: childrenByParentId,
                showExecutingState: true
            ) == .running {
                return true
            }
        } else if child.output == nil || hasRunningDescendant(nestedChildren, childrenByParentId: childrenByParentId) {
            return true
        }
    }
    return false
}

private func agentStatuses(from payload: [String: Any]?) -> [String] {
    guard let payload else { return [] }
    let states = payload["agentsStates"] as? [String: Any] ?? payload["agents_states"] as? [String: Any]
    guard let states else { return [] }
    return states.values.compactMap { value in
        guard let status = (value as? [String: Any])?["status"] as? String else { return nil }
        return normalizeStatus(status)
    }
}

private func toolOutputPayload(_ output: String) -> [String: Any]? {
    if let direct = jsonObject(from: output) {
        return direct
    }

    // Codex collab results arrive as text content blocks whose text is a JSON payload.
    guard let array = jsonValue(from: output) as? [[String: Any]] else { return nil }
    for block in array {
        guard let text = block["text"] as? String, let payload = jsonObject(from: text) else { continue }
        return payload
    }
    return nil
}

private func jsonObject(from string: String) -> [String: Any]? {
    parsedToolInputObject(string)
}

private func jsonValue(from string: String) -> Any? {
    guard let data = string.data(using: .utf8) else { return nil }
    return try? JSONSerialization.jsonObject(with: data)
}

private func stringValue(_ value: Any?) -> String? {
    value as? String
}

private func normalizeStatus(_ status: String) -> String {
    status.replacingOccurrences(of: "_", with: "").lowercased()
}

// MARK: - Sub-Agent Metadata
//
// Port of `frontend/src/lib/sub-agent.ts` (`parseSubAgentInfo`, `buildChildrenMap`).

struct SubAgentInfo {
    let subagentType: String
    let description: String
    let prompt: String?
    let runInBackground: Bool
    let model: String?
    let tool: String?
}

/// Parse sub-agent metadata from a Task/Agent tool's input JSON.
func parseSubAgentInfo(_ tool: ToolCall) -> SubAgentInfo? {
    guard tool.name == "Task" || tool.name == "Agent" else { return nil }
    guard let input = jsonObject(from: tool.input) else { return nil }
    return SubAgentInfo(
        subagentType: (input["subagent_type"] as? String) ?? "Agent",
        description: (input["description"] as? String) ?? "",
        prompt: input["prompt"] as? String,
        runInBackground: (input["run_in_background"] as? Bool) == true,
        model: input["model"] as? String,
        tool: input["tool"] as? String
    )
}

/// Build a map of parentToolUseId → children for hierarchical rendering.
func buildChildrenMap(_ tools: [ToolCall]) -> [String: [ToolCall]] {
    var map: [String: [ToolCall]] = [:]
    for tool in tools {
        guard let parent = tool.parentToolUseId else { continue }
        map[parent, default: []].append(tool)
    }
    return map
}

File: frontend/src/lib/sub-agent-status.ts
import type { ToolCall } from "@/types";
import { parseSubAgentInfo } from "@/lib/sub-agent";

export type SubAgentExecutionState = "pending" | "running" | "completed" | "failed";

type JsonObject = Record<string, unknown>;

const RUNNING_AGENT_STATUSES = new Set(["pendinginit", "running", "interrupted"]);
const FAILED_AGENT_STATUSES = new Set(["errored", "notfound", "failed"]);

export interface SubAgentExecutionOptions {
  showExecutingState?: boolean;
  children?: ToolCall[];
  childrenMap?: Map<string, ToolCall[]>;
}

export function getSubAgentExecutionState(
  tool: ToolCall,
  options: SubAgentExecutionOptions = {},
): SubAgentExecutionState {
  if (tool.name !== "Task" && tool.name !== "Agent") {
    return tool.output === undefined ? "pending" : "completed";
  }

  const info = parseSubAgentInfo(tool);
  const inputPayload = parseJsonObject(tool.input);
  const outputPayload = parseToolOutputPayload(tool.output);
  const toolStatus = normalizeStatus(stringValue(outputPayload?.status ?? inputPayload?.status));
  const agentStatuses = [
    ...agentStatusesFromPayload(inputPayload),
    ...agentStatusesFromPayload(outputPayload),
  ];

  if (toolStatus === "failed" || agentStatuses.some((status) => FAILED_AGENT_STATUSES.has(status))) {
    return "failed";
  }

  if (!options.showExecutingState) {
    return tool.output === undefined ? "pending" : "completed";
  }

  if (tool.output === undefined) {
    return "running";
  }

  if (
    info?.tool === "spawnAgent" &&
    agentStatuses.some((status) => RUNNING_AGENT_STATUSES.has(status))
  ) {
    return "running";
  }

  if (hasRunningDescendant(options.children ?? [], options.childrenMap)) {
    return "running";
  }

  return "completed";
}

export function isSubAgentRunning(tool: ToolCall, options: SubAgentExecutionOptions = {}): boolean {
  return getSubAgentExecutionState(tool, options) === "running";
}

function hasRunningDescendant(children: ToolCall[], childrenMap: Map<string, ToolCall[]> | undefined): boolean {
  for (const child of children) {
    const nestedChildren = childrenMap?.get(child.id) ?? [];
    if (child.name === "Task" || child.name === "Agent") {
      if (getSubAgentExecutionState(child, {
        showExecutingState: true,
        children: nestedChildren,
        childrenMap,
      }) === "running") {
        return true;
      }
      continue;
    }
    if (child.output === undefined || hasRunningDescendant(nestedChildren, childrenMap)) {
      return true;
    }
  }
  return false;
}

function agentStatusesFromPayload(payload: JsonObject | null): string[] {
  const states = asObject(payload?.agentsStates) ?? asObject(payload?.agents_states);
  if (!states) return [];
  return Object.values(states)
    .map((entry) => normalizeStatus(stringValue(asObject(entry)?.status)))
    .filter((status): status is string => status !== undefined);
}

function parseToolOutputPayload(output: string | undefined): JsonObject | null {
  if (!output) return null;

  const parsed = parseJson(output);
  const direct = asObject(parsed);
  if (direct) return direct;

  // Codex collab results arrive as text content blocks whose text is a JSON payload.
  if (!Array.isArray(parsed)) return null;
  for (const block of parsed) {
    const text = stringValue(asObject(block)?.text);
    const payload = text ? parseJsonObject(text) : null;
    if (payload) return payload;
  }
  return null;
}

function parseJsonObject(value: string | undefined): JsonObject | null {
  return asObject(parseJson(value));
}

function parseJson(value: string | undefined): unknown {
  if (!value) return null;
  try {
    return JSON.parse(value) as unknown;
  } catch {
    return null;
  }
}

function asObject(value: unknown): JsonObject | null {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as JsonObject
    : null;
}

function stringValue(value: unknown): string | undefined {
  return typeof value === "string" ? value : undefined;
}

function normalizeStatus(value: string | undefined): string | undefined {
  if (!value) return undefined;
  return value.replace(/_/g, "").toLowerCase();
}
